sound/soc/codecs/rt715-sdca.c

```diff
@@ -88,171 +88,338 @@ static int rt715_sdca_set_amp_gain_put(struct snd_kcontrol *kcontrol,
 	struct soc_mixer_control *mc =
 		(struct soc_mixer_control *)kcontrol->private_value;
 	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
-	unsigned int val_l, val_r, gain_l_val, gain_r_val, loop_cnt, i, reg;
+	unsigned int val_l, val_r, gain_val[2] = {0};
 	int ret;
 
-	if (strstr(ucontrol->id.name, "FU02 Capture Volume") ||
-		strstr(ucontrol->id.name, "FU06 Capture Volume"))
-		loop_cnt = 2;
-	else if (strstr(ucontrol->id.name, "FU0E Boost") ||
-		strstr(ucontrol->id.name, "FU0C Boost"))
-		loop_cnt = 4;
-	else
-		loop_cnt = 1;
-
 	/* control value to 2s complement */
 	/* L channel */
-	gain_l_val = ucontrol->value.integer.value[0];
-	if (gain_l_val > mc->max)
-		gain_l_val = mc->max;
-	val_l = gain_l_val;
-
-	if (mc->shift == 8) {
-		gain_l_val = (gain_l_val * 10) << mc->shift;
-	} else {
-		gain_l_val =
-			((abs(gain_l_val - mc->shift) * RT715_SDCA_DB_STEP) << 8) / 1000;
-		if (val_l <= mc->shift) {
-			gain_l_val = ~gain_l_val;
-			gain_l_val += 1;
-		}
-		gain_l_val &= 0xffff;
+	gain_val[0] = ucontrol->value.integer.value[0];
+	if (gain_val[0] > mc->max)
+		gain_val[0] = mc->max;
+	val_l = gain_val[0];
+
+	gain_val[0] =
+		((abs(gain_val[0] - mc->shift) * RT715_SDCA_DB_STEP) << 8)
+			/ 1000;
+	if (val_l <= mc->shift) {
+		gain_val[0] = ~gain_val[0];
+		gain_val[0] += 1;
 	}
-
+		gain_val[0] &= 0xffff;
 	/* R channel */
-	gain_r_val = ucontrol->value.integer.value[1];
-	if (gain_r_val > mc->max)
-		gain_r_val = mc->max;
-	val_r = gain_r_val;
-
-	if (mc->shift == 8) {
-		gain_r_val = (gain_r_val * 10) << mc->shift;
-	} else {
-		gain_r_val =
-			((abs(gain_r_val - mc->shift) * RT715_SDCA_DB_STEP) << 8) / 1000;
-		if (val_r <= mc->shift) {
-			gain_r_val = ~gain_r_val;
-			gain_r_val += 1;
-		}
-		gain_r_val &= 0xffff;
+	gain_val[1] = ucontrol->value.integer.value[1];
+	if (gain_val[1] > mc->max)
+		gain_val[1] = mc->max;
+	val_r = gain_val[1];
+
+	gain_val[1] =
+		((abs(gain_val[1] - mc->shift) * RT715_SDCA_DB_STEP)
+			<< 8) / 1000;
+	if (val_r <= mc->shift) {
+		gain_val[1] = ~gain_val[1];
+		gain_val[1] += 1;
 	}
+	gain_val[1] &= 0xffff;
 
 	/* Lch*/
+	ret = regmap_write(rt715->mbq_regmap, mc->reg, gain_val[0]);
+	if (ret != 0) {
+		dev_err(component->dev, "Failed to write 0x%x=0x%x\n",
+			mc->reg, gain_val[0]);
+		return ret;
+	}
+	/* Rch */
+	ret = regmap_write(rt715->mbq_regmap, mc->rreg, gain_val[1]);
+	if (ret != 0) {
+		dev_err(component->dev, "Failed to write 0x%x=0x%x\n",
+			mc->rreg, gain_val[1]);
+		return ret;
+	}
```

Comment:
the two channel processing code fragments look identical to me, with the only difference being `mc->reg` vs. `mc->rreg`. So if you put these into an array, you could make that a loop? Like
```
for (i = 0; i < 2; i++) {
	gain_val = ucontrol->value.integer.value[i];
	...
	ret = regmap_write(rt715->mbq_regmap, reg[i], gain_val);
	if (ret < 0) {
		...
	}
}
```